debug: tracing: Add Segger RTT linker section options

giansta opened the pull request:
Allows optionally placing Segger RTT data either in a specific
linker section that is located at RAM start, or in a specific linker
section defined by a memory region in DTS, as third and fourth
alternative to the DTCM section or the default data section.
This is useful to share the fixed address for different programs,
typically bootloader and application, and have seamless logging.

Signed-off-by: Giancarlo Stasi <giancarlo.stasi.co@gmail.com>

fixes #53544

giansta commented:
Expected to fix https://github.com/zephyrproject-rtos/zephyr/issues/53544 with https://github.com/zephyrproject-rtos/segger/pull/17 in Segger repo

microbuilder commented:
@nordic-krch Seems OK to me, but mind having a look since you maintain the Segger integration?

giansta commented:
> If you mean that the approval depends on the approval of the segger PR, I totally agree :-)

giansta commented:
> @nordic-krch Seems OK to me, but mind having a look since you maintain the Segger integration?

If you mean that the approval depends on the approval of the segger PR, I totally agree :-)

giansta commented:
Changed after comments in https://github.com/zephyrproject-rtos/zephyr/issues/53544.
I believe that the first compliance error happened because I pushed the project segger after this push.
The 2nd and 3rd errors, on the other hand, look docker internal issues.

giansta committed:
debug: tracing: Add Segger RTT linker section options

Allows optionally placing Segger RTT data either in a specific
linker section that is located at RAM start, or in a specific linker
section defined by a memory region in DTS, as third and fourth
alternative to the DTCM section or the default data section.
This is useful to share the fixed address for different programs,
typically bootloader and application, and have seamless logging.

Signed-off-by: Giancarlo Stasi <giancarlo.stasi.co@gmail.com>

nordic-krch commented on drivers/debug/Kconfig.rtt:
```diff
@@ -74,6 +74,9 @@ config SEGGER_RTT_SECTION_NONE
 	bool "Place RTT data in the DTCM linker section"
 
+config SEGGER_RTT_SECTION_FIXED_ADDR
+	bool "Place RTT data in specific linker section with fixed address"
```
don't we want `default y`? If bootloader is not used then it has no difference and when bootload is used then probably most users would like to get logs from bootloader and application.

giansta replied:
Agree, done.

giansta replied:
I've added a `if CPU_CORTEX_M`, thanks.

nordic-krch reviewed (commented):
Can you also update west so that it points to the changes in segger repo?

giansta commented:
> 



> Can you also update west so that it points to the changes in segger repo?

I already tried, but I got a compliance error with CI and I had to remove it. I'll try again.

giansta commented:
> Can you also update west so that it points to the changes in segger repo?

It should be done now.

nordicjm commented:
@nordic-krch can you re-review this please?

carlescufi commented on west.yml:
```diff
@@ -209,7 +209,8 @@ manifest:
       revision: 2097f26ce7dee36922b22ce048c09cd073ebae3d
     - name: segger
-      revision: e2ff2200556e8a8f962921444275c04971a2bb3d
+      url: https://github.com/giansta/segger.git
```
Please remove this, you need to send your change to the Zephyr repo instead

giansta replied:
Removed that commit. I expect I have to update `west.yml` only once that the segger repo PR  #17 is merged, right?

giansta committed:
drivers: console: rtt_console: use Segger recommended write API

Use recommended write API that internally locks RTT usage and checks
if the RTTcontrol block initialization is done.

Signed-off-by: Giancarlo Stasi <giancarlo.stasi.co@gmail.com>

carlocaione commented on include/zephyr/arch/arm/aarch32/cortex_m/scripts/linker.ld:
```diff
@@ -232,6 +232,16 @@ SECTIONS
 	. = ALIGN(_region_min_align);
 	_image_ram_start = .;
 
+#if defined(CONFIG_SEGGER_RTT_SECTION_FIXED_ADDR)
+	SECTION_DATA_PROLOGUE(_RTT_SECTION_NAME,(NOLOAD),)
+	{
+	__rtt_buff_data_start = .;
+	*(.rtt_buff_data)
+	__rtt_buff_data_end = ALIGN(4);
+	} GROUP_DATA_LINK_IN(RAMABLE_REGION, RAMABLE_REGION)
+	__rtt_buff_data_size = __rtt_buff_data_end - __rtt_buff_data_start;
+#endif
+
```
What about using `zephyr_linker_sources_ifdef()` instead?

giansta replied:
Isn't it a CMake function? Indeed I see it used only in CMakeLists.txt files, while linker file don't use it. Can it be used in linker files?

carlocaione replied:
No, it is used by CMake to embed linker script snippets.

giansta replied:
I've never done that way. On the other end, this linker script has similar sections with conditional compile switch, it was natural for me to do the same. Do you have an example?

giansta replied:
Would it be possible to separate this request in another PR?

sigvartmh replied:
Here is how I did it when trying to implement the same as you have done here: https://github.com/sigvartmh/fw-nrfconnect-zephyr-1/commit/009b403829368e7a6687aa46a24d63f22d71a8b1#diff-7ab5276f0606a2bd1620a91984e170d371c28c4d7884774234ebb7340936ec06

sigvartmh replied:
Sorry I forgot to include the segger_rtt.ld in my dump of this :/ it is pretty much the same as the section you have there.

giansta replied:
Thanks for the suggestions and sorry for my delay. It should be done now.

MarcianoPreciado commented:
We're so close on this, any chance these changes can get made?

MarekPieta commented on include/zephyr/arch/arm/aarch32/cortex_m/scripts/linker.ld:
```diff
@@ -232,6 +232,16 @@ SECTIONS
 	. = ALIGN(_region_min_align);
 	_image_ram_start = .;
 
+#if defined(CONFIG_SEGGER_RTT_SECTION_FIXED_ADDR)
```
That could be helpful also if a bootloader does not enable RTT at all (bootloader still may accidentally modify RAM area which application uses for the RTT). For the bootloader use-case, maybe we could allow to reserve memory without actually enabling the RTT support?

MarekPieta replied:
I also wonder if we could somehow share required size of the RTT related section among bootloader and application (configuration must be aligned to ensure interoperability, otherwise we may see some strange errors). Maybe we could somehow propagate the information e.g. using SysBuild already

sigvartmh replied:
I'm not sure Zephyr's build system + linker have anyway of putting it at the end of RAM? 

One way of achieving this is to ensure that the RTT section is the last thing placed in RAM as then we could use some of the values of the KConfig to figure out the size of the partition then set the (Location Counter)[https://ftp.gnu.org/old-gnu/Manuals/ld-2.9.1/html_chapter/ld_3.html#SEC10] to be  `RTT_START_ADDR = RAM SIZE - SIZE of RTT Block.`

The reason why this must be placed last is that if we move the LC forward we cannot move it back in GNUs LD. 

> Side note: ARMs scatter file/ARMs linker does have this ability.

sigvartmh replied:
> That could be helpful also if a bootloader does not enable RTT at all (bootloader still may accidentally modify RAM area which application uses for the RTT). Alternatively, for the mentioned bootloader use-case, maybe we could allow to reserve memory without actually enabling the RTT support?

Agreed this would be useful ref. https://interrupt.memfault.com/blog/noinit-memory#retrofitting-a-noinit-region-when-you-cant-update-the-bootloader

nordicjm replied:
It would just be better to have a memory region setup in the dts and use that specifically for RTT. It should not just be assumed to be at the end of RAM, there are many reasons why it should not be placed there.

sigvartmh replied:
@nordicjm : Sure a separate linker memory section for RTT would work fine thanks for the input 👍 Do you know how this can be achieved to section off the RAM by any chance?  Is there a way to define a node that would cut off the end of ram. 

IMO it should be user selectable where you want to place your RTT section but at least having it BEFORE and AFTER Application/Bootloader RAM would be two sensible options.

@giansta : https://docs.zephyrproject.org/apidoc/latest/devicetree__regions_8h.html Whatever Docs I found on this.

giansta replied:
I didn't want to add this at beginning, as the case for this PR (and for related issue #53544) was having a bootloader and app sharing the configuration.
But yes, the case with another bootloader where we want to keep the application logs between resets is also interesting. I've added a 2nd custom behaviour to use a DTS memory region that can be placed at end of RAM or at another address. I've tested it defining a region for my board. The two different choices are necessary, because of the snippet usage for the former one.
Anyway I wouldn't set it as default choice, because it needs a DTS definition, I kept the "normal" start of RAM option as default.

sigvartmh commented on drivers/console/rtt_console.c:
```diff
@@ -41,9 +41,7 @@ static int rtt_console_out(int character)
-		SEGGER_RTT_LOCK();
-		cnt = SEGGER_RTT_WriteNoLock(0, &c, 1);
-		SEGGER_RTT_UNLOCK();
+		cnt = SEGGER_RTT_Write(0, &c, 1);
```
For another PR but could we make the channel configurable?

giansta replied:
Even simple, I agree to do it in another PR :-)

giansta committed:
debug: tracing: Add Segger RTT init mode configuration

Allows RTT inizialization function to either init Cntrol Block always
or initialize only after checking it it's not already initialized by
another program, typically by a bootloader.

Signed-off-by: Giancarlo Stasi <giancarlo.stasi.co@gmail.com>

giansta commented:
Hello. Even if this is a low prio PR, I did some job on it time ago trying to address all requests.
Also, the [corresponding PR](https://github.com/zephyrproject-rtos/segger/pull/17) in segger repo is still open.
Is it possible to reopen and review it? I don't seem to have the rights to reopen it.

nordicjm commented:
@giansta the segger PR needs rebasing

giansta commented:
> @giansta the segger PR needs rebasing

Rebased, thanks!

nordicjm commented:
@ioannisg @bbolen @microbuilder @povergoing @nordic-krch @MarekPieta since this is a very old PR and gets merge conflicts can you review and approve if you find it acceptable so that the module update can be merged and this is able to be merged without encountering another merge conflict?

kartben commented:
@nordic-krch would be good to try to get this and the associated Segger PR in for 3.7?

ithinuel commented on modules/segger/Kconfig:
```diff
@@ -82,6 +83,43 @@ config SEGGER_RTT_SECTION_DTCM
+	bool "Place RTT data in custom linker section at RAM start"
+
+config SEGGER_RTT_SECTION_CUSTOM_DTS_REGION
+	bool "Place RTT data in custom linker section defined by a memory region in DTS"
```
I'm not fully familiar with devicetree & its use in Zephyr.

Where is the link with this config and the DTS configuration. Would you have an example for it please?

nordicjm replied:
Ping @giansta

giansta replied:
This feature was requested [here](https://github.com/zephyrproject-rtos/zephyr/pull/53569#discussion_r1247899876).
The link is just the config `SEGGER_RTT_SECTION_CUSTOM_NAME` that is enabled both using `SEGGER_RTT_SECTION_CUSTOM` and `SEGGER_RTT_SECTION_CUSTOM_DTS_REGION`. It's used in `SEGGER_RTT_Conf.h` in the modules repo.

Defining the DTS is then system or board specific. For my nRF52 board I tested it defining in DTS a region like this:
```
/ {
	soc {
		sram_rtt: memory@20000800 {
			compatible = "zephyr,memory-region", "mmio-sram";
			zephyr,memory-region = "SRAM_RTT";
			reg = <0x2000F800 0x800>;
		};
	};
};
```
The section name is defined in the line `zephyr,memory-region = "SRAM_RTT";` and both application and MCUboot prj.conf had to define `CONFIG_SEGGER_RTT_SECTION_CUSTOM_NAME="SRAM_RTT"`.

cfriedt reviewed (approved):
Definite improvement

cfriedt commented:
Currently building against `v3.6-branch`, I'm getting a weird warning 

```shell
warning: orphan section `.rtt_buff_data' from `modules/segger/libmodules__segger.a(SEGGER_RTT.c.obj)' being placed in section `.rtt_buff_data'
```

Was this addressed already? I thought I remember seeing something about it.

cfriedt commented:
Was able to verify this working with a custom stm32 board, `tests/boot/test_mcuboot`

Also needed the wrapper [here](https://github.com/fab-uleuh/zephyr-twister-rtt), which would also be a good reason to build something like this into west.

There is a PR for that.

carlescufi commented:
@giansta thanks so much for your patience.

giansta commented:
Thanks @carlescufi! Anyway, this PR requires even https://github.com/zephyrproject-rtos/segger/pull/17
I would expect it to be merged too.

gchwier commented:
Hi, after merging this PR not working tests:
* https://github.com/zephyrproject-rtos/zephyr/tree/main/tests/subsys/dfu/img_util

on nrf52840dk/nrf52840.
To reproduce:
`${ZEPHYR_BASE}/scripts/twister -p nrf52840dk/nrf52840 -T ${ZEPHYR_BASE}/tests/subsys/dfu/img_util -s dfu.image_util  --west-flash --device-testing --device-serial /dev/ttyACM0 -c -vv`
output:
```
DEBUG   - DEVICE: Running TESTSUITE img_util
DEBUG   - DEVICE: ===================================================================
DEBUG   - DEVICE: START - test_check_flash
DEBUG   - DEVICE: ASSERTION FAIL [0] @ WEST_TOPDIR/zephyr/lib/utils/onoff.c:283
DEBUG   - DEVICE: E: r0/a1:  0x00000004  r1/a2:  0x0000011b  r2/a3:  0x00000001
DEBUG   - DEVICE: E: r3/a4:  0x00000004 r12/ip:  0x00000040 r14/lr:  0x00000871
DEBUG   - DEVICE: E:  xpsr:  0x21000000
DEBUG   - DEVICE: E: Faulting instruction address (r15/pc): 0x000074d0
DEBUG   - DEVICE: E: >>> ZEPHYR FATAL ERROR 4: Kernel panic on CPU 0
DEBUG   - DEVICE: E: Current thread: 0 (unknown)
DEBUG   - DEVICE: E: Halting system
DEBUG   - Timed out while monitoring 
```

bisect shows that commit: https://github.com/zephyrproject-rtos/zephyr/commit/711ed082a2c361311a06b2c76c604f7e03fa8ddf

kartben commented:
@gchwier I think it might be because https://github.com/zephyrproject-rtos/segger/pull/17 still hasn't been merged? cc @carlescufi

gchwier commented:
@kartben I've used that PR and looks that it does not help, still the same issue.

brycewilkins commented:
@giansta I hope you don't mind me trying to help get your Segger repo PR merged! I've encountered the problems here also and would love to see this resolved. Great work!
